File: 684_Redundant_Connection.cpp
class Solution {
public:
    int find(int p, vector<int>& parent)
    {
        if (p == parent[p]) return p;
        return parent[p] = find(parent[p], parent);
    }
    bool unite(int u, int v, vector<int>& parent)
    {
        int pu = find(u, parent);
        int pv = find(v, parent);

        if (pu == pv)
            return true;
        parent[pv] = pu;
        return false;
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();           

        vector<int> parent(n + 1);
        for (int i = 1; i <= n; ++i)
            parent[i] = i;

        vector<int> ret(2, -1);
        for (auto e : edges)
        {
            int u = e[0], v = e[1];
            if (unite(u, v, parent);
            { 
                ret[0] = u;
                ret[1] = v;
            }
        }

        return ret;
    }
};

File: 684_Redundant_Connection.py
class Solution(object):
    def findRedundantConnection(self, edges):
        """
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        n = len(edges)
        parent = [i for i in range(n + 1)]

        ret = [-1, -1]

        for edge in edges:
            u, v = edge[0], edge[1]
            if self.unite(u, v, parent):
                ret = [u, v]

        return ret

    def find(self, v, parent):
        if v == parent[v]:
            return v
        parent[v] = self.find(parent[v], parent)
        return parent[v]

    def unite(self, u, v, parent):
        pu = self.find(u, parent)
        pv = self.find(v, parent)

        if pu == pv:
            return True

        parent[pu] = pv
        return False



